llvm/lib/CodeGen/LiveVariables.cpp

```diff
@@ -277,11 +277,13 @@ void LiveVariables::HandlePhysRegUse(Register Reg, MachineInstr &MI) {
           continue;
         if (PartDefRegs.count(SubReg))
           continue;
+
+        // Check if SubReg is defined at LastPartialDef.
+        bool IsDefinedHere = LastPartialDef->modifiesRegister(SubReg, TRI);
         // This part of Reg was defined before the last partial def. It's killed
         // here.
-        LastPartialDef->addOperand(MachineOperand::CreateReg(SubReg,
-                                                             false/*IsDef*/,
-                                                             true/*IsImp*/));
+        LastPartialDef->addOperand(
+            MachineOperand::CreateReg(SubReg, IsDefinedHere, true /*IsImp*/));
```

Comment:
Swapping what used to be a use with a def sounds wrong. Do we need multiple implicit operands to model this liveness?

Reply:
It will only be switched if the use is defined at this line. If the use is omitted or marked as `implicit-def`, the test passes.

Multiple implicit operands will be added if they are subregs of `Reg` within `LiveVariables::HandlePhysRegUse`, so I believe it is already being modeled.

Reply:
> we just need to remove the implicit use.

So can you just skip the add operand altogether? I'm still confused about why this pass is doing anything. This is very old code that predates subregister tracking. I bet we can delete all of this code as a step towards deleting the entire pass

Reply:
Yes, it can be skipped. However, `llvm/test/CodeGen/AArch64/ldrpre-ldr-merge.mir` would need to be updated to not mark `implicit-def $w2_hi`, which may be incorrect, so I believe it is better to leave the `implicit-def` in.

I'd argue that we are not really replacing what used to be a use with a def. They should have been defs all along because the sub or super register has a definition at this line.

If we do not fix it here, the logic to fix it in LiveRangeCalc.cpp needs to update `LiveRangeCalc::findReachingDefs` to take into consideration definitions that are found at the current instruction. And it propagates a change to be made in RegisterPressure.cpp, so I think it is cleaner to have it here.

Reply:
LiveRangeCalc requiring this pass run before is spaghetti. We shouldn't need this pass to fix the liveness representation. The liveness of physical registers should be fully expressed at all points in the compilation. This pass has been papering over issues in LiveRangeCalc 

The implicit-def of w2_hi is redundant with the x2 implicit-def. The use was marking a kill (except it didn't add a kill flag?). So there's no reason to have the new implicit-def, and losing the implicit use loses the use kill point

Reply:
Removed a chunk of the code, which also got rid of `w2_hi`, so now it doesn't have a kill point... Need to take a closer look at it to see if we can keep `w2_hi` here.